### PR TITLE
Middleware: Add CSP Report Only support

### PR DESCRIPTION
**Pull request by jcalisto**
**What is this feature?**

Adds a new `content_security_policy_report_only` config to the `security` section, to allow setting a `Content-Security-Policy-Report-Only` in the middleware. 

A [Content-Security-Policy-Report-Only](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) allows to experiment a policy and monitor the effects without enforcing it.

**Which issue(s) does this PR fix?**:
https://github.com/grafana/hosted-grafana/issues/2681

It will help us investigating the impact of enabling the CSP on cloud instances, and can also be useful for other users trying to experiment with their own CSP. 

**Special notes for your reviewer**:
There was a need to refactor the CSP middleware function as there is common behaviour between the CSP and CSP-Report-Only headers (i.e., policy nonce and policy variables).

### CODE REVIEW

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 1]
```diff
@@ -621,6 +621,15 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
 
 Set Content Security Policy template used when adding the Content-Security-Policy header to your requests. `$NONCE` in the template includes a random nonce.
 
+### content_security_policy_report_only
+
+Set to `true` to add the Content-Security-Policy-Report-Only header to your requests. CSP in Report Only mode allows to experiment with policies by monitoring their effects without enforcing.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:626]:
```suggestion
Set to `true` to add the Content-Security-Policy-Report-Only header to your requests. CSP in Report Only mode enables you to experiment with policies by monitoring their effects without enforcing them.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -621,6 +621,15 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
...
+Both policies can be enabled simultaneously.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:627]:
```suggestion
You can enable both policies simultaneously.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 3]
```diff
@@ -621,6 +621,15 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
...
+
+### content_security_policy_template
+
+Set Content Security Policy Report Only template used when adding the Content-Security-Policy-Report-Only header to your requests. `$NONCE` in the template includes a random nonce.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:631]: I'm not sure what you are trying to say here. Is it: "Set the Content Security Policy Report Only template that is used when you add the Content-Security-Policy-Report-Only header to your requests."?

**Review by chri2547 — COMMENTED** [review]
One suggestion and one question. Thank you!

File: docs/sources/setup-grafana/configure-grafana/_index.md (hunk 3 above)
Comment (jcalisto) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: @chri2547 I updated the wording of this phrase, hopefully it makes more sense now.

Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md]: 👍

**Review by chri2547 — COMMENTED** [review]
The docs look fine to me. I'm looking for a tech review approval from @papagian , @sh0rez , or @idafurjes .

File: conf/sample.ini [hunk 4]
```diff
@@ -327,6 +327,16 @@
...
+# Enable adding the Content-Security-Policy-Report-Only header to your requests.
+# Allows you to monitor the effects of a policy without enforcing it.
+;content_security_policy_report_only = false
+
+# Set Content Security Policy Report Only template used when adding the Content-Security-Policy-Report-Only header to your requests.
+# $NONCE in the template includes a random nonce.
+# $ROOT_PATH is server.root_url without the protocol.
+;content_security_policy_report_only_template = """script-src 'self' 'unsafe-eval' 'unsafe-inline' 'strict-dynamic' $NONCE;object-src 'none';font-src 'self';style-src 'self' 'unsafe-inline' blob:;img-src * data:;base-uri 'self';connect-src 'self' grafana.com ws://$ROOT_PATH wss://$ROOT_PATH;manifest-src 'self';media-src 'none';form-action 'self';"""
+
+
```
Comment (hairyhenderson) [inline conf/sample.ini:339]: extra newline can be removed here

```suggestion

```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -619,7 +619,16 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
 
 ### content_security_policy_template
 
-Set Content Security Policy template used when adding the Content-Security-Policy header to your requests. `$NONCE` in the template includes a random nonce.
+Set the policy template that will be used when adding the Content-Security-Policy header to your requests. `$NONCE` in the template includes a random nonce.
```
Comment (hairyhenderson) [inline docs/sources/setup-grafana/configure-grafana/_index.md:622]:
```suggestion
Set the policy template that will be used when adding the `Content-Security-Policy` header to your requests. `$NONCE` in the template includes a random nonce.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 6]
```diff
@@ -619,7 +619,16 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
...
+
+### content_security_policy_report_only
+
+Set to `true` to add the Content-Security-Policy-Report-Only header to your requests. CSP in Report Only mode enables you to experiment with policies by monitoring their effects without enforcing them.
```
Comment (hairyhenderson) [inline docs/sources/setup-grafana/configure-grafana/_index.md:626]:
```suggestion
Set to `true` to add the `Content-Security-Policy-Report-Only` header to your requests. CSP in Report Only mode enables you to experiment with policies by monitoring their effects without enforcing them.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 7]
```diff
@@ -619,7 +619,16 @@ Set to `true` to add the Content-Security-Policy header to your requests. CSP al
...
+You can enable both policies simultaneously.
+
+### content_security_policy_template
+
+Set the policy template that will be used when adding the Content-Security-Policy-Report-Only header to your requests. `$NONCE` in the template includes a random nonce.
```
Comment (hairyhenderson) [inline docs/sources/setup-grafana/configure-grafana/_index.md:631]:
```suggestion
Set the policy template that will be used when adding the `Content-Security-Policy-Report-Only` header to your requests. `$NONCE` in the template includes a random nonce.
```

File: pkg/middleware/csp.go [hunk 8]
```diff
@@ -14,40 +14,64 @@ import (
 	"github.com/grafana/grafana/pkg/setting"
 )
 
-// AddCSPHeader adds the Content Security Policy header.
-func AddCSPHeader(cfg *setting.Cfg, logger log.Logger) func(http.Handler) http.Handler {
+// ContentSecurityPolicy adds the Content-Security-Policy header.
```
Comment (hairyhenderson) [inline pkg/middleware/csp.go:17]:
```suggestion
// ContentSecurityPolicy sets the configured Content-Security-Policy and/or Content-Security-Policy-Report-Only header(s) in the response.
```

**Comment by hairyhenderson** [comment]
Overall this LGTM @jcalisto 👍

I'd like to see a review from @grafana/backend-platform too though! 😉

**Review by sakjur — APPROVED** [review]
Code changes look great, glad to see the consideration put into these changes and improving the general state around the code you changed :)

I haven’t tested or compared the implementation to the spec, fyi

**Comment by jcalisto** [comment]
Credits to @hairyhenderson for his great suggestions!

**Pull request merged**